### PR TITLE
HPCC-17461 Rationalise the use of TimeElapsed

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [x] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browers
  - [ ] The component(s) render as expected

## Testing:

Minimal testing so far on a small sample of queries.

### CODE REVIEW

**Comment by ghalliday** [comment]
@GordonSmith please take a look at this branch.  I think it should provide pairs for all graphs/subgraphs in hthor/roxie and thor.  It also switches compiler times to be total rather than elapsed.

**Comment by GordonSmith** [comment]
@ghalliday Will I ever see the a scope with all three of the following attributes:
* Started
* Finished
* TimeElapsed

Also do I consider it an error if I see a TimeElapsed without a Started?

Finally can I consider a Started + Finshed to be equal to a Started + TimeElapsed?

**Comment by ghalliday** [comment]
Yes, there will be some scopes will all 3 defined.

In the long term there should always be a WhenStarted if there is a TimeElapsed.  One exception may be activities - which have a TimeFirstRow instead.  Short term you are likely to find quite a few which no corresponding WhenXXX

WhenStarted + TimeElapsed should equal WhenFinished.  However they are recorded at different degrees of accuracy: TimeElapsed is more accurate than WhenFinished-WhenStarted.

**Comment by ghalliday** [comment]
See also https://github.com/hpcc-systems/LN/pull/838

**Comment by richardkchapman** [comment]
Checklist is missing

**Comment by ghalliday** [comment]
@richardkchapman now ready for review.  Rebased onto latest 6.4 and conflicts fixed.

File: ecl/hthor/hthor.cpp [hunk 1]
```diff
@@ -236,8 +236,8 @@ void CHThorActivityBase::updateProgressForOther(IStatisticGatherer &progress, un
 {
     StatsEdgeScope scope(progress, otherActivity, whichOutput);
     progress.addStatistic(StNumRowsProcessed, numProcessed);
-    progress.addStatistic(StNumStarted, 1);  // wrong for an activity in a subquery
-    progress.addStatistic(StNumStopped, 1);
+    progress.addStatistic(StNumStarts, 1);  // wrong for an activity in a subquery
```
Comment (richardkchapman) [inline ecl/hthor/hthor.cpp]: Is this comment still valid?

**Comment by richardkchapman** [comment]
@ghalliday Needs to translate old names in existing workunits to new names. Should push back to 7.0.0 to get that done.

File: ecl/hthor/hthor.cpp (hunk 1 above)
Comment (ghalliday) [inline ecl/hthor/hthor.cpp]: Yes - it isn't keeping track of the number of times the child query has been started/stopped.

**Comment by ghalliday** [comment]
@GordonSmith do you specially look out for NumStarted on a graph?  Otherwise I can't think of any problems with renaming it - it isn't stored as text in a workunit.

@richardkchapman I have switched this back to 6.4 - this is a safer version.  Rather than reusing the previous enum I have added new values, which means the old names will be there.

**Comment by ghalliday** [comment]
(Reasons for including in 6.4 is that it consistently adds start/end times to graphs/subgraphs.  But I am happy to rebase to master if preferred or still too much potential for problems.)  Main problem I can think of is TimeElapsed replaced with TimeLocalExecute for timings from eclccserver.

**Comment by GordonSmith** [comment]
@ghalliday I do special case "NumStarted" and "NumStopped" in js and c++ GraphControl

**Comment by ghalliday** [comment]
ok.  I'll retarget to master, and open a couple of jiras to ensure the graph control spots the new ones

**Comment by ghalliday** [comment]
@richardkchapman rebased, and a jira updated/isue opened on graph control.

**Pull request merged**